lib/android.js

```diff
@@ -1498,54 +1659,534 @@ function unwrapMethodId (methodId) {
+
+    // Restore core args, callee-saves & LR.
+    writer.putPopRegReg('x29', 'lr');
+    writer.putPopRegReg('x27', 'x28');
+    writer.putPopRegReg('x25', 'x26');
+    writer.putPopRegReg('x23', 'x24');
+    writer.putPopRegReg('x21', 'x22');
+    writer.putPopRegReg('x7', 'x20');
+    writer.putPopRegReg('x5', 'x6');
+    writer.putPopRegReg('x3', 'x4');
+    writer.putPopRegReg('x1', 'x2');
+
+    // Restore FPRs.
+    writer.putPopRegReg('d6', 'd7');
+    writer.putPopRegReg('d4', 'd5');
+    writer.putPopRegReg('d2', 'd3');
+    writer.putPopRegReg('d0', 'd1');
+
+    writer.putBCondLabel('ne', 'invoke_replacement');
+
+    do {
+      offset = relocator.readOne();
+    } while (offset < ART_QUICK_CODE_HOOK_REDIRECT_SIZE_ARM64);
+
+    relocator.writeAll();
+
+    if (!relocator.eoi) {
+      writer.putBranchAddress(target.add(offset));
+    }
+
+    writer.putLabel('invoke_replacement');
+
+    writer.putLdrRegRegOffset('x16', 'x0', artMethodOffsets.quickCode);
+    writer.putBrReg('x16');
+
+    writer.flush();
+  });
+
+  return offset;
+}
+
+const artQuickCodePrologueWriters = {
+  ia32: writeArtQuickCodePrologueX86,
+  x64: writeArtQuickCodePrologueX86,
+  arm: writeArtQuickCodePrologueArm,
+  arm64: writeArtQuickCodePrologueArm64
+};
+
+function writeArtQuickCodePrologueX86 (target, trampoline) {
+  Memory.patchCode(target, 16, code => {
+    const writer = new X86Writer(code, { pc: target });
+
+    writer.putJmpAddress(trampoline);
+    writer.flush();
+  });
+}
+
+function writeArtQuickCodePrologueArm (target, trampoline) {
+  const targetAddress = target.and(THUMB_BIT_REMOVAL_MASK);
+
+  Memory.patchCode(targetAddress, 16, code => {
+    const writer = new ThumbWriter(code, { pc: targetAddress });
+
+    writer.putLdrRegAddress('pc', trampoline.or(1));
+    writer.flush();
+  });
+}
+
+function writeArtQuickCodePrologueArm64 (target, trampoline) {
+  Memory.patchCode(target, 16, code => {
+    const writer = new Arm64Writer(code, { pc: target });
+
+    writer.putBranchAddress(trampoline);
+    writer.flush();
+  });
+}
+
+class ArtQuickCodeInterceptor {
+  constructor (quickCode) {
+    this.quickCode = quickCode;
+    this.trampoline = null;
+    this.overwrittenPrologue = null;
+    this.overwrittenPrologueLength = 0;
+  }
+
+  _createTrampoline () {
+    if (trampolineAllocator === null) {
+      const trampolineSize = (pointerSize === 4) ? 128 : 256;
+      trampolineAllocator = makeCodeAllocator(trampolineSize);
+    }
+
+    this.trampoline = trampolineAllocator.allocateSlice();
+  }
+
+  activate (vm) {
+    this._createTrampoline();
+
+    const writeTrampoline = artQuickCodeReplacementTrampolineWriters[Process.arch];
+    const prologueLength = writeTrampoline(this.trampoline, this.quickCode, vm);
+    this.overwrittenPrologueLength = prologueLength;
+
+    const quickCodeAddress = (Process.arch === 'arm')
+      ? this.quickCode.and(THUMB_BIT_REMOVAL_MASK)
+      : this.quickCode;
```

Comment:
```suggestion
        ? this.quickCode.and(THUMB_BIT_REMOVAL_MASK)
        : this.quickCode;
```

Though semistandard might disagree – if it does then we'll go with what it dictates. (Did you run semistandard on this PR btw?)

Reply:
It's applied now.